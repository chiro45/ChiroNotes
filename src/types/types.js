
//definimos los tipos que vamos a ver en el state del redux

export const types ={
    //types login
    login: '[auth] Login',
    logout: '[auth] Logout',

    //uiErrors
    uiSetError: '[UI] Set Error',
    uiRemoverError: '[UI] Remove Error ',
    //loadings
    uiStartLoading: '[UI] Start Logading',
    uiFinishLoading: '[UI] Finish Loading',
    //notes
    notesAddNew: '[Notes] New Note',
    notesActive: '[Notes] Active Note',
    notesLoad: '[Notes] Load Notes',
    notesUpdated: '[Notes] Updated Note',
    notesFileUrl: '[Notes] Update Image url',
    notesDelete: '[Notes] Delete',
    notesLogoutCleaning: '[Notes] Logout Cleaning'

}
    
