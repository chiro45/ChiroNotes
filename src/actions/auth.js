import Swal from 'sweetalert2'


import { types } from "../types/types"
import {firebase, googleAuthProvider} from '../firebase/firebaseConfing'
import { finishLoading, startLoading } from "./ui"
import { noteLogout} from './notes'
//acciones que son usadas en el auth




export const startLoginEmailPassword = (email,password)=>{
    return(dispatch)=>{
        
        dispatch(startLoading())
        firebase.auth().signInWithEmailAndPassword(email,password)
        .then( ({ user }) =>{


            dispatch(login(user.uid, user.displayName))
            
            
        dispatch(finishLoading())

        })
        .catch(e => {
        Swal.fire("Error","El usuario o contraseÃ±a son incorrectos","error")
    })
}}
export const startRegisterWhithEmailPasswordName = (email,password, name)=>{
    return(dispatch)=>{
        firebase.auth().createUserWithEmailAndPassword(email,password)
        .then( async({ user })=>{

            await user.updateProfile({
                displayName: name

            })

            dispatch(
                login(user.uid, user.displayName)
             )
            
         
        })
        .catch(e => {
            Swal.fire("Error","El usuario ya ha sido registrado","error")
        
    })

}}


export const startGoogleLogin = ()=>{
    return (dispatch)=>{
       firebase.auth().signInWithPopup(googleAuthProvider) 
       .then(({ user })=>{

           dispatch(
               login(user.uid, user.displayName, user.photoURL),

               console.log(user.photoURL)
                 
               )
            })
            
       
       
       
    }
}


export const login = (uid, displayName, photoURL)=>{
        return{
            type: types.login,
            payload: {
                uid,
                photoURL,
                displayName,
            }
        }
}


export const startLogout = () =>{
    return async(dispatch)=>{
        firebase.auth().signOut()

        dispatch(logout())
        
        dispatch(noteLogout())
    }
}

export const logout = () =>({
    type: types.logout
})


