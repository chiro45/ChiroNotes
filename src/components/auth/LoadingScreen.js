import React from 'react';

export const LoadingScreen = () => {
  return (
    <div className='auth__main'>
    <div className='auth__box-container'>
        <h2 className='auth__loading-title'>Cargando <i className=" spinner fas fa-spinner"></i></h2>
    </div>
    </div>
  );
};
