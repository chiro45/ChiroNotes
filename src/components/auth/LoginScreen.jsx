import  { useState } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { Link } from 'react-router-dom'
import { startGoogleLogin, startLoginEmailPassword } from '../../actions/auth'
import { useForm } from '../../hooks/useForm';
import validator  from 'validator';
import { removeError, setError } from '../../actions/ui';



export const LoginScreen = () => {
                   
    const [show, setShow] = useState(false);                            //uso el state para mostrar la contraseña o no
    const {msgError} = useSelector(state=>state.ui)                     //usamos el hook use selector para extraer el msgError del store
    const dispatch = useDispatch();                                     //extraemos el useDispatch


    const [formValues, handleInputChange] = useForm({                   //coustomHook para el manejo del formulario
        email: '',
        password:''
    });

    const {email, password} = formValues;                               //desestructuracion de los datos
    
    const isFormValid = ()=>{                                           //validacion del formulario
        if(!validator.isEmail(email)){
            dispatch(setError('El email es incorrecto'))
            return false;
        }else if( password.length < 8){
            dispatch(setError('La contraseña es menor a 8 caracteres o estas no coinciden'))
            return false;
        }
        dispatch(removeError())
        
        return true;
    }

    const handleLogin = (e)=>{                                           //realiza el envio del formulario, si tiene errores los muestra y sino realiza el dispatch
       e.preventDefault();
       if(isFormValid()){
        dispatch(startLoginEmailPassword(email, password))

       }
    }

    const handleGoogleLogin = ()=>{                                      //en el caso del login con google nos permite realizarla
        dispatch(startGoogleLogin())
    }
    const showOrHiddenPassword = ()=> setShow(!show);                    //funcion para mostrar o no la contraseña

    return (
       
        
        <div>
        
            <h3 className='auth__title'>ChiroNotes <i className=" far fa-edit"></i></h3>
        
        <form onSubmit={handleLogin} className={`animate__animated animate__fadeIn animate__faster`}>
            {    
              msgError &&                                           //realiza la operacion logica para la muestra o no de los errores
             
             <div className='auth__alert-error'>
             <p>{`${msgError}`}</p>
             </div>
                 
             }
        
        <input
        className='auth__input'
        type='text'
        placeholder='Email'
        name='email'
        autoComplete='off'
        value={email}
        onChange={handleInputChange}
        />
        <input
        className='auth__input'
        type={show ? 'text' : 'password'}
        placeholder='Contraseña'
        name='password'
        value={password}
        onChange={handleInputChange}
        />
        <div className='auth__checkbox'>
        <p className='auth__show-passwd'>Mostrar contraseña</p>
        <input type='checkbox'
         onChange={showOrHiddenPassword}/>
        </div>

        <button
        type='submit'
        className='btn btn-primary btn-block'
      
        >Ingresar <i className=' icons fas fa-arrow-right'></i>
        </button>
        
        <div className='auth__social-networks'>
            <p><b></b></p>
             
        <div className="google-btn" 
        onClick={handleGoogleLogin}
        >
            <div className="google-icon-wrapper">
                <img className="google-icon" src="https://upload.wikimedia.org/wikipedia/commons/5/53/Google_%22G%22_Logo.svg" alt="google button" />
            </div>
                 <p className="btn-text">
                <b>Ingresa con google</b>
                     </p>
        </div>
        </div>
        <Link to='/auth/register'
        className='link'>
           Crear una cuenta nueva <i className="icons fa fa-angle-double-right"></i>
        </Link>

        </form>

             
    </div>
    
   
    )
}
