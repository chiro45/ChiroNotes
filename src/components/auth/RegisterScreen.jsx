import React from 'react'
import { Link } from 'react-router-dom'
import { useForm } from '../../hooks/useForm';
import validator  from 'validator';
import {startRegisterWhithEmailPasswordName} from '../../actions/auth';
import { useDispatch, useSelector } from 'react-redux';
import { removeError, setError } from '../../actions/ui';




export const RegisterScreen = () => {
    const dispatch = useDispatch();
    const {msgError} = useSelector(state=>state.ui)

    const [formValues, handleInputChange] = useForm({           //utilizamos el coustomHook para el manejo del formulario 
        name: 'fernando',                                               //declaramos el initialState
        email: 'fernando@gmail,cim',
        password: '123123',
        password2: '123123',
    });

    const {name, email , password, password2} = formValues;     //desestructuracion de los datos
    const handleRegister = (e)=>{
        e.preventDefault();
        if(isFormValid()){
            dispatch(startRegisterWhithEmailPasswordName(email, password, name)) //si no hay errores en el form se realiza la operacion

        }
    }

    
    const isFormValid = ()=>{                                                           //realiza la verificacion de los datos
        if(name.trim().length < 5 ){
            dispatch(setError('El nombre de usuario debe ser mayor a 5 caracteres'))
            return false;
        } else if(!validator.isEmail(email)){
            dispatch(setError('El email es incorrecto'))
            return false;
        }else if(password !== password2 || password.length < 8){
            dispatch(setError('La contrase単a es menor a 8 caracteres o estas no coinciden'))
            return false;
        }
        dispatch(removeError())
        return true;
    }
    return (
        <div>
            <h3 className='auth__title'>ChiroNotes <i className="  far fa-edit"></i></h3>
                
            
            <form
             onSubmit={handleRegister}
             className={`animate__animated animate__fadeIn animate__faster`}>

                 {    
                  msgError &&                   //realiza la comprobacion para mostrar en pantalla los errores o no
                 
                 <div className='auth__alert-error'>    
                 <p>{`${msgError}`}</p>
                 </div>
                     
                 }
            <input
            className='auth__input'
            type='text'
            placeholder='Nombre'
            name='name'
            autoComplete='off'
            onChange={handleInputChange}
            />
            <input
            className='auth__input'
            type='text'
            placeholder='Email'
            name='email'
            autoComplete='off'
            onChange={handleInputChange}
            />
            <input
            className='auth__input'
            type='text'
            placeholder='Contrase単a'
            name='password'
            onChange={handleInputChange}
            />
            <input
            className='auth__input'
            type='text'
            placeholder='Confirma tu Contrase単a'
            name='password2'
            onChange={handleInputChange}
            />

            <button
            type='submit'
            className='btn btn-primary btn-block marginBottom-5'
           
            >Registrarse <i className=' icons fa fa-sign-in-alt'></i>
            </button>
            
            
            <Link to='/auth/login'
            className='link '>
              多Ya te has Registrado?
            </Link>

            </form>


        </div>
    )
}
