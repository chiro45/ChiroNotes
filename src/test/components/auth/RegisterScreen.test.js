import React from 'react';
import { mount } from 'enzyme';
import { Provider } from 'react-redux'
import { MemoryRouter } from 'react-router-dom';

import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import '@testing-library/jest-dom';
import { RegisterScreen } from '../../../components/auth/RegisterScreen';
import { types } from '../../../types/types';







const middlewares = [thunk];
const mockStore = configureStore(middlewares);


const initState = {
    auth: {},
    ui: {
        loading: false,
        msgError: null
    }
};

let store = mockStore(initState);



const wrapper = mount( 
    <Provider store={ store }>
        <MemoryRouter>
            <RegisterScreen /> 
        </MemoryRouter>
    </Provider>

)
describe('Pruebas en <RegisterScreen/>', () => {
    
    beforeEach(()=>{
        store = mockStore(initState)
        jest.clearAllMocks();
    })

    test('debe de mostrarase correctamente', () => {
       expect(wrapper).toMatchSnapshot();
    });


    test('debe de hacer el dispatch de la accion respectiva', () => {
      
        const emailField = wrapper.find('input[name="email"]')
    
        emailField.simulate('change',{
            target:{
                value:'lucianodavid',
                name: 'email'
            }
        })

        wrapper.find('form').prop('onSubmit')({ 
            preventDefault(){}
      });

        const actions = store.getActions();
        // expect(actions).toEqual({
        //     type: types.uiSetError,
        //     payload: 'El nombre de usuario debe ser mayor a 5 caracteres'
        // })
        

    });


    test('debe de mostrar la caja de alerta con el error', () => {
        const initState = {
            auth: {},
            ui: {
                loading: false,
                msgError: 'Email no es correcto'
            }
        };
        
        let store = mockStore(initState);
        
        
        
        const wrapper = mount( 
            <Provider store={ store }>
                <MemoryRouter>
                    <RegisterScreen /> 
                </MemoryRouter>
            </Provider>
        
        )

       expect( wrapper.find('.auth__alert-error').exists()).toBe(true)

       expect( wrapper.find('.auth__alert-error').text().trim()).toBe(initState.ui.msgError)


    });
    
    
    
    

});