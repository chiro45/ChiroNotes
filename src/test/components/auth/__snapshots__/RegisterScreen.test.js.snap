// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <RegisterScreen/> debe de mostrarase correctamente 1`] = `
<div>
  <h3
    className="auth__title"
  >
    ChiroNotes 
    <i
      className="  far fa-edit"
    />
  </h3>
  <form
    className="animate__animated animate__fadeIn animate__faster"
    onSubmit={[Function]}
  >
    <input
      autoComplete="off"
      className="auth__input"
      name="name"
      onChange={[Function]}
      placeholder="Nombre"
      type="text"
    />
    <input
      autoComplete="off"
      className="auth__input"
      name="email"
      onChange={[Function]}
      placeholder="Email"
      type="text"
    />
    <input
      className="auth__input"
      name="password"
      onChange={[Function]}
      placeholder="Contraseña"
      type="text"
    />
    <input
      className="auth__input"
      name="password2"
      onChange={[Function]}
      placeholder="Confirma tu Contraseña"
      type="text"
    />
    <button
      className="btn btn-primary btn-block marginBottom-5"
      type="submit"
    >
      Registrarse 
      <i
        className=" icons fa fa-sign-in-alt"
      />
    </button>
    <a
      className="link "
      href="/auth/login"
      onClick={[Function]}
    >
      ¿Ya te has Registrado?
    </a>
  </form>
</div>
`;
